Add clean datasource feature

jrschumacher opened the pull request:
Add the helper to clean the datasource before running the next commands.

- Had to use beforeEach because of the `whenCalledRemotely` being called `beforeEach`

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

ritch commented on lib/helpers.js:
```diff
@@ -26,6 +26,17 @@ _beforeEach.withApp = function(app) {
+    if(!dsName) dsName = 'db';
+
+    if (this.app) {
+      this.app.datasources[dsName].automigrate();
```
Ensure `automigrate()` completes... As we have it here it will run alongside the tests...

jrschumacher replied:
Thanks for the catch, didn't think about that. Originally this was after which theoretically would work, but still fail on the next describe...

ritch replied:
@jrschumacher do you have time to fix this? If not I can create a pull request and fix this in order to land this.

You should be able to just add a callback to `beforeEach`.

jrschumacher replied:
@ritch thanks for following up. I forgot about that todo.

So your saying like this?

```js
_beforeEach.cleanDatasource = function(dsName, cb) {
  beforeEach(function() {
    //...
    cb();
```

bajtos replied:
Nope, the following is correct:

```js
_beforeEach.cleanDatasource = function(dsName){
  beforeEach(function(done) {
    //...
    done();
```

jrschumacher replied:
@ritch @bajtos I added it can you check. The test fails now though. Will need to examine that.

bajtos commented:
@slnode ok to test

bajtos commented:
It makes me wonder, why the model cleanup installed by `givenModel` is not enough? ([code](https://github.com/strongloop/loopback-testing/blob/f836e19c560b71bae837748780d477142a438d9b/lib/helpers.js#L106-L108))

jrschumacher commented:
I had no idea that was there. I'll provide an example of the issue I was
running into.

On Monday, January 19, 2015, Miroslav Bajtoš <notifications@github.com>
wrote:

> It makes me wonder, why the model cleanup installed by givenModel is not
> enough? (code
> <https://github.com/strongloop/loopback-testing/blob/f836e19c560b71bae837748780d477142a438d9b/lib/helpers.js#L106-L108>
> )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/strongloop/loopback-testing/pull/40#issuecomment-70537011>
> .
>


-- 
Ryan Schumacher

twitter / @jrschumacher <http://twitter.com/jrschumacher>

jrschumacher commented:
@bajtos could you look at e97141ab? This is why I want to be able to clean datasources. In my new describe I would like the ability to have a clean database and the `givenModel` isn't cleaning up after itself -- or at lease not how I need it.

bajtos commented:
I guess the problem is that `givenModel` cleans the database in `afterEach` hook, while it should clean it in a `beforeEach` hook.

As I am thinking about this more, also in the context of https://github.com/strongloop/loopback-example-full-stack/pull/53/, it's probably best to introduce this new helper method as you proposed.

Reworking `givenModel` to clean the database in a `beforeEach` hook would be tricky: first of all, it is a breaking change. Secondly, the code installing the cleanup hook would have to be clever so that the second call of `givenModel` does not remove the model created by the first call.

jrschumacher commented:
@bajtos @ritch I fixed the test. I added a conditional check [here](https://github.com/strongloop/loopback-testing/pull/40/files#diff-27a1c86133de4310373decfdad200a91R33) which previously broke please make sure this is valid.

Also let me know what the commit message should be and I'll squash it down.

ritch commented:
> Also let me know what the commit message should be and I'll squash it down.

Not necessary
